### ISSUE TITLE
On Containers In Serverless

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#460]
There are a few, large architectural problems that we need to solve in the Framework and introducing “Containers” is an unusual, yet intriguing solution...  Here is a proposal on how Containers could solve some of our biggest issues:

### How It Could Work

We could put a “containers” folder in place of the “back” folder, and the "containers" folder would contain…  containers.

```
project
   containers
      container1
      container2
```
**Containers would have 2 purposes:**
- To encapsulate code that shares a runtime (nodes, python, java, etc.) 
- To encapsulate code that you want to share across multiple Lambda functions.  Currently in the framework, we have a simple list of modules and there is a lot of hacking as to how code is shared across them.  Bringing on multi-language support will only complicate this.  In this proposed “container” version of the framework, code is shared at the container-level only.

Since each container has a designated runtime, this will allow us to more easily support bringing multiple languages into the framework.
```
project
   containers
      restApi     // Node.js
      imgresizers // Python
```
Since each container also serves a designated code-sharing level, people should be allowed to make multiple containers with the same runtimes:
```
project
   containers
      restApi            // Node.js
      eventHandlers    // Node.js
      scheduledFunctions  // Python
      imgResizers     // Python
```      
Serverless will only package and zip up individual containers for upload to AWS Lambda, so you can put a lib folder in your containers:

```
project
   containers
      nodejs_restApi
         lib
         module1
         module2
```
**Serverless Dependency Management**

A large problem we currently have in Serverless is dependency management.  When each Serverless Module has their own set of dependencies, the result is messy and laden with issues.  Many users are requesting a single set of dependencies shared by all Serverless Modules.  

With containers, sharing language-specific dependencies across modules seems easier to implement.
```
project
   containers
      nodejs_restApi
         package.json
         lib
         module1
```
**Serverless Module Architecture & Delivery**

We’re interested in creating a robust Serverless Module ecosystem (because nothing is cooler than sharing pre-written Lambda code) however we haven’t been able to make that work because of the complexity involved.

Since containers make it easier to share dependencies across Serverless Modules, we should reconsider using language-specific package managers as the delivery mechanism for Serverless Modules, our Modules can be easily installed at the container root.

We can also revert back to our previous Module architecture, which contains both a language-specific module and Serverless Module:

```
// NPM example
serverless_npm_module
   lib
   s
     serverless-users-crud
        s-module.json
        functionOne
```
On installation, Serverless Framework can optionally add everything in `s` folder to your project.  The `lib` is requirable into your code like any other project.  In other words, adding Serverless/Lambda scaffolding is option on top of a normal language-specific module.

What do you think of this approach and what other benefits could containers bring?

**Comment by ac360** [comment]
We could also change the name from "containers" to something else.  Some ideas that have come up:

* components
* services
* apps

**Comment by visualasparagus** [comment]
+1 for 'containers' as this relates more to its purpose of separating runtimes (a la docker containers). The terms 'components', 'services' and 'apps' are confusing to me as it gets muddled into the 'modules' and 'functions' terminology.

**Comment by Nopik** [comment]
The concept seems interesting, at least at first glance. It will be very usefyl, if executed right. I'm kind of hating how the current module folder structure look like, the this proposal seems much cleaner and usable.

I guess that majority of the projects will end up with 1 container, though. If that is the case, I would look for the ways to drop down 1 folder hierarchy level (e.g. reduce `containers/my_container` to something like `my_container` or something like this). But, maybe it wont be good to have such optimization.

What would be the structure inside container? Namely, can we go with no enforced structure inside? I would love to not be forced to create extra depth for modules within containers. I mean, my lambdas are already organized into some folders, according to the business logic, but forcing stuff like `modules/` or `lib/` or `back/modules/` inside container (let alone stuff like enforcing `back/modules/my_module/lib/`) would be really annoying for me, and basically defeat the purpose of simplifying stuff with containers.

**Comment by eahefnawy** [comment]
hmmm interesting! So that means we'll have both **Serverless Containers** and **Serverless Modules** inside those containers? If yes, I think this will lead to confusion + additional boilerplate. Also would that mean both containers and modules are sharable/installable? That's even more confusing.

If you mean we'll only have containers which have lambda functions inside, then how are those containers different from the current module implementation that we have now? It seems like just a different name. We could just encourage code sharing only in module-level the same way we would encourage code sharing only in container level.

Overall, since our issue is allowing users to share code across modules, I don't think adding an extra level of directory structure would solve that, because what if someone wants to share code across containers the same way some people want to share code across modules. If we'll make it a best practice not to share code between containers, then why not just make it a best practice not to share code across modules?!

Just my two cents :)

**Comment by marklawlor** [comment]
I've put some thought into containers and I've come to the conclusion that they are probably a bad idea...mostly for the reasons @eahefnawy mentioned. I liked it at first, but they don't actually solve the original problem discussed in #428 

I believe that root of this issue is that Serverless is trying to support multiple projects, with different runtimes and dependancies, all that the same time. All for the sole purpose of deployment to AWS.

My idea is largely inspired via Docker Swarm, which allows you to take a pool of Docker hosts can combine them into 1 virtual host. It also borrows heavily from `npm link`.

Firstly drop the idea of projects and containers. Lets just pretend we only have modules. Modules have:

- A single runtime (node/python/etc)
- A single set of dependancies
- Have one/many endpoints
- Have one/many lambdas

To group two modules together we use the CLI to 'cluster' them. 

A new command should be created:  `serverless link <cluster-name>`.  When run it creates a folder in `~/.serverless/<cluster-name>` with a symlink to the project (pretty much exactly what `npm link` does)

When running `serverless dash` and other Serverless commands, the lambdas/endpoints from your other clustered/linked projects will also appear and can be actioned.

When searching for a project in the cluster, Serverless first checks `npm_modules/ruby bundles/etc` before looking for the project in `~/.serverless/<cluster-name>/`. This allows 3rd party modules to be added to the cluster just like normal dependancies.

A second command:  `serverless shrinkwrap <cluster-name>` will also need to be created. This command creates a clusterfile.json that contains the names/urls of all projects in the cluster. This allows you to check clusters into your versioning.

TLDR:  Drop the idea of projects with multiple modules, only have individual modules. Allow modules to be cluster/grouped for deployment in a similar fashion to `npm link`. 

I know its a big shift, but I hopefully my rambling makes sense and could be an alternative solution. I really like this approach as it drops the old school 'monolithic project' problem and adopts for a more modern modular approach. It also bring in a much lower barrier to entry for new developers and allows projects to grow organically and in a workflow that is similar to modern node development.

**Comment by zevmo** [comment]
We all realize that the way to create, maintain, and deploy these various languages, resources, dependancies, and releases will be challenging. I'm in agreement on the Docker Swarm-esque approach. I also agree that the manifesto of microservices, whilst remaining flexible enough to accommodate all types of apps (large and small), should be the goal of the architecture.

My only concern is on the shifting sands of the several core pieces underlying the structure of this local development workflow. I'm not saying it isn't an inherent risk in any of these projects, but it certainly is something to keep in mind, since Docker itself has an ongoing refactoring that is highly volatile.

But I think the approach is good, and the difficulty is in the orchestration, and simplicity of setup. This will work to bolster the decision for developer, so they find it easy enough to adopt Serverless as their chosen process of creating awesome apps on AWS.

**Comment by ryansb** [comment]
## Naming things

I think the name "containers" is going to end up super confusing - especially if anyone ever has a project that involves Lambda and [ECS](https://aws.amazon.com/ecs/). If we do go with some higher-than-module grouping, I'd much rather use "cluster" or "team" or "group" or something. 

I think trying to build in *more* modularity might be excessive in terms of the return for actual developers. I'm not saying this because concerns above are invalid, but because those concerns are already well-handled by language package managers. 

## Project structure 

Also, it would mean we have a heirarchy of `project/back/{app or container}/{module}/{function}/{actual files}`, and I'm already not big on the level of nesting we have. 

As of the most recent v0.1.0 branch, here's what we have:

```
├── admin.env
├── back
│   └── modules
├── meta
│   ├── private
│   │   ├── resources
│   │   │   └── s-resources-cf-development-useast1.json
│   │   └── variables
│   │       ├── s-variables-common.json
│   │       ├── s-variables-development.json
│   │       └── s-variables-development-useast1.json
│   └── public
│       └── variables
│           └── s-variables-common.json
├── plugins
│   └── custom
├── README.md
```

If we're going to make a big change to the modules, I think we could stand to kill off `back` entirely and have `modules` be top-level. 

## Modularity and reuse

I think everyone agrees we aren't trying to be the "we do NodeJS and API Gateway and that's all" framework, so I'm going to take that for granted for the rest of this. The use cases that can already exist are (pardon me if I leave some out)

1. REST API. API-Gateway, Lambda, etc to hook up Lambda to HTTP things
1. Eventful workers. SNS, S3, or other event source hooked up in some kind of producer/consumer pipeline
1. Cloud Cronjobs *(tm)*. Scheduled events to do backups, check compliance, or a zillion other things
1. Invokeable Backends. No API Gateway, functions get invoked through the Lambda InvokeAsync or InvokeSync API calls from an SDK on the client (major for mobile).

Because of this range, we have to think about what we *can* build that will support (or at least not break) all of these use cases. The jobs that we can do that fall into that group include:

1. Making deployable zip files with the function's dependencies
1. Grouping functions into some unit where they can share code easily
1. Uploading code and deploying it with Lambda
1. Managing stages from dev -> test -> production (or whatever number of stages users want)
1. Allowing different configs between environments (think 12-factor app)
1. Creating resources functions need to work

And most importantly: make developers productive at making $THING by letting them use as much existing knowledge as possible. 

In python, you can have git (or hg, or tarball) URLs in your requirements file to install things from private repos, nodejs has the same thing through npm, and so does Java. That's a perfectly fine way to share code that's sufficiently reusable to be a (language) module. Because lots of very smart and hardworking folks have *already built this*, I don't think we should re-do it in `serverless`. 

Having groups of functions confined to modules makes a lot of sense to me, and I could see @marklawlor 's argument in favor of clusters/teams. One missing bit (IMO) is dependencies between modules. Let's say I'm building an app with user/auth functions, news feed functions, and search functions. The latter two need to use user/auth functionality to do their jobs (show the right news feed, and search only things that user has permissions for). Clustering would need to support this dep graph

```
users <----o newsfeed
        \--o search
```

Newsfeed can't run without having the user/auth functions deployed in its environment. What these *don't* do is share code between each other, they only communicate via HTTP or direct invokes, and if you have some sort of language client for `users` you would make a `npm` or `pypi` module that would be in the `package.json` or `requirements.txt` and installed on deploy in the search/newsfeed zipfiles. 

## Actionable Ideas

I think the module should remain the boundary for:

1. Runtime selection
1. Code sharing
1. Deployment (so functions can count on all being deployed together inside a module, with the same module libraries)
1. Logical functionality (this module does X)
1. All things that are language-specific (package.json, requirements.txt, and so forth)

The module doesn't need to be a boundary for:

1. Language-packaged code sharing (python or nodejs packages)
1. Environment variables
1. IAM roles and other resources (CFN templates)

I think these boundaries, plus a group/team feature like @marklawlor suggested would be the best way to move ahead because we would support just about every Lambda use case, and support the ability for people to share serverless modules (ecosystems are the best!).

**Comment by kevinold** [comment]
I agree with others that the name "containers" is confusing.  I believe it would require an explanation in a FAQ somewhere and could be a distraction.

I'm wondering if we should borrow from existing, familiar terminology.

"module" seems to be well known and what's proposed is a grouping of these, I'd favor simply calling it "modules".   Another idea would be "packages", since from the Node community we already have the terminology and working system of a "package" housing a "module" (or "modules" to be exported).

Related to sharing code, I would not expect my containers to explicitly be able to share internals (i.e. require something from container1 in container2).  This should be handled by something outside of the container.  I think "libs", "libraries" would fit this use case.

**Comment by remicastaing** [comment]
You can find some inspiration [here](http://www.thesaurus.com/browse/container) for a suiting terminology. Why not  "cauldron"? ;)

**Comment by visualasparagus** [comment]
@ryansb, I really like your writeup of the issues. Here's my 2¢ about a few things to add to the discussion. I might be going out on a limb here, but please bear with me.

### Naming things
After reading the comments, I agree that 'container' isn't the right choice. However, I also don't really like 'components', 'services' or 'apps' either. I do like @kevinold's suggestion of 'package'.

### Project structure
I have a few ideas I'd like to throw out there. On the whole the idea was to simplify and flatten using a few different organizational concepts (module, runtime and function). Each one also takes into consideration the following.
1. The nesting is getting pretty deep. What if we were to kill both 'back' and 'modules'?
2. Why not combine s-project.json into a custom package.json?
3. I don't actually understand why the plugins folder is part of the project and not part of the serverless install. It seems redundant to install them for every project instead of having them in a central location and then just reference them in the s-project.json's `plugins` attribute.
4. Since I'm on the subject, why not remove `plugins` and just have `node_modules`? A slss command `slss install plugin` could be created that wraps `npm install --save` and then also adds it to the 'serverless.json' plugins section?
5. Should we not separate the API Gateway endpoint definitions from the function definition? If I want to use someone's function code, I don't really want to be bound by their API Gateway configuration.
6. Then change `meta` to `.meta` or `.slss` or something to the directory namespace clean.


##### module above all
this seems to the more or less the direction of the current conversation regarding generally where things go.
````
project/
	admin.env
    package.json	-- formerly s-project.json. includes plugins that aren't related to the specific lambda runtime, but just slss
    node_modules	-- essentially a project specific plugins dir
    api_endpoints.json	-- project wide endpoint to function mapping.
 	module_one_nodejs/
		module.json	-- formerly s-module.json, it mainly determines runtime
		package.json	-- node dependencies if runtime is nodejs. would include at least serverless-helpers-js
		node_modules	-- node dependency dir if nodejs
		lib/	-- module wide code common
		functions/
			function_one/
				function.json	-- formerly s-function.json. excludes endpoints attribute.
				event.json
				handler.js
 	module_two_python/
		module.json	-- formerly s-module.json, it mainly determines runtime
		requirements.txt	-- python dependencies if runtime is python. would include at least python version of serverless-helpers-js.
		local_dependency_dir_for_python	-- i'm not sure about this?
		lib/	-- module wide code common
		functions/
			function.json	-- formerly s-function.json. excludes endpoints attribute.
			event.json
			handler.py
	.meta/	-- cf resources, variables, etc
````


##### runtime above all
if we have to group things by runtime, then why not make that division part of the project structure?
````
project/
	admin.env
    package.json	-- formerly s-project.json. includes plugins that aren't related to the specific lambda runtime, but just slss
    node_modules	-- essentially a project specific plugins dir
    api_endpoints.json	-- project wide endpoint to function mapping.
    node_js/ -- runtime specific directory
 		lib/	-- runtime wide code common
		module_one/
			module.json	-- formerly s-module.json, it mainly determines runtime
			package.json	-- node dependencies if runtime is nodejs. would include at least serverless-helpers-js
			node_modules	-- node dependency dir if nodejs
			lib/	-- module wide code common
			functions/
				function_one/
					function.json	-- formerly s-function.json. excludes endpoints attribute. 
					event.json
					handler.js
 	python/ -- runtime specific directory
 		lib/	-- runtime wide code common
		module_one/
			module.json	-- formerly s-module.json, it mainly determines runtime
			requirements.txt	-- python dependencies if runtime is python. would include at least python version of serverless-helpers-js.
			local_dependency_dir_for_python	-- i'm not sure about this?
			lib/	-- module wide code common
			functions/
				function.json	-- formerly s-function.json. excludes endpoints attribute.
				event.json
				handler.py
	.meta/	-- cf resources, variables, etc
````


##### function above all
since the only technical limitation is that the function needs a runtime, why not this?
````
project/
	admin.env
    package.json	-- formerly s-project.json. includes plugins that aren't related to the specific lambda runtime, but just slss
    node_modules	-- essentially a project specific plugins dir
    api_endpoints.json	-- project wide endpoint to function mapping
    node_js_lib/
	   	package.json	-- common module dependencies in one spot
	   	node_modules	-- common module. would include at least serverless-helpers-js
		lib/	-- module wide code common for node js
    python_lib/
		requirements.txt	-- python dependencies if runtime is python. would include at least python version of serverless-helpers-js.
		local_dependency_dir_for_python	-- i'm not sure about this?
		lib/	-- module wide code common for python
    module_one/ -- logical directory. there is no longer a module.json
		function_one_nodejs/
			function.json	-- formerly s-function.json. excludes endpoints attribute, includes runtime and module.json cloudFormation attribute. 
			package.json	-- node dependencies if runtime is nodejs.
			node_modules	-- node dependency dir if nodejs
			lib/	-- function specific code
			event.json
			handler.js
 		function_one_python/
			function.json	-- formerly s-function.json. excludes endpoints attribute, includes runtime and module.json cloudFormation attribute. 
			requirements.txt	-- python dependencies if runtime is python. would include at least python version of serverless-helpers-js.
			local_dependency_dir_for_python	-- i'm not sure about this?
			lib/	-- function specific code
			event.json
			handler.py
	.meta/	-- cf resources, variables, etc
````

### Reuse
I can imagine, I might want to install only one function from a re-usable module.
````
slss install module modulename --all
slss install module modulename functionnameone functionnametwo
````

### Conclusion
There are likely pros and cons to all of these and I'm not aware of all the previous discussions that have been had. I'm putting this out there as a point of conversation. I'm interested to know any thoughts.

**Comment by ac360** [comment]
Thank you all for your feedback.  We've drafted out a few Architectural Proposals here:  https://docs.google.com/document/d/1JBWTKsVaxnnQAwHyxtubhj5z4pS0xjJZkXg4L1o0xrU

**Comment by visualasparagus** [comment]
1. Shouldn't the level of code sharing (so module?) be the (standard) location for dependency management to help promote module reuse and distribution?

2. Is there ever a use case for having different functions with different runtimes in one module? if so, shouldn't the runtime be specified in the s-function.json? Dependencies could still be managed at the module level though and just include both a package.json and a requirements.txt.

3. If i install someone else's module/function, this will include the s-function.json. Should I be bound by their decisions regarding the endpoint and events definitions? What if this collides with another module I've installed? What if I don't want to expose their functions as endpoints or have different event sources for the same functionality? Won't updating the module at a later date overwrite any of my changes if we're using npm or pip for dependency management?

4. serverless plugins are also just npm modules, correct? shouldn't these just be managed with npm as well and stored in a project level package.json and 'node_modules' directory?

**Comment by ac360** [comment]
Implemented in `v0.1`, except they are called Components.  Thank you everyone for the wonderful suggestions!
